### PR TITLE
Only rebuild scenarios on a relevant file change

### LINKED ISSUE #608: Only build scenarios on a change and push to a release
Currently, the scenarios are re-built on every run of the CI workflows. This is unnecessary as the binaries themselves don't change that often, just new scenarios are added.

The idea of this issue is to build all the scenarios on a release and then push them as artefacts to that release.

If the scenarios do change as part of a PR then they do need to be rebuilt and the new binaries need to be tested. If possible, it would be nice to only rebuild the affected scenarios and use the prebuilt binaries of the unaffected scenarios.

AC:
- [ ] Upon a release, all the scenarios are built and uploaded as release artefacts
- [ ] Upon a push to `main`, create a new pre-release with the new scenario binary ZIP files - should be some cleanup around old pre-releases.
- [ ] In CI, if any of the scenario, Wind Tunnel, or Zome code changes then the scenarios are rebuilt and the new scenarios are used to test.
  - (_Stretch goal_) Only the affected scenarios are rebuilt

### PR DESCRIPTION
**Pull request by cdunster**
### Option 1 of 2

> [!Note]
> There is an alternative version of this PR (#619) that uses the Nix input derivation hash to check for changes.

> [!Warning]
> Only one of this or #619 should be merged.

### Summary

This PR follows on from #618, and causes the scenarios to only be re-built, uploaded to the S3 bucket, and used by the Nomad nodes if the files relevant to that scenario have changed. Otherwise, the scenario zip files from the latest release or pre-release are used. It is possible to force a rebuild when running manually.

Must be merged after #618.

Closes #608.

Things that should trigger a rebuild of the scenarios:
```
├── bindings
│   ├── client ✔ (just HC ones)
│   ├── kitsune_client ✔ (just K2 ones)
│   ├── kitsune_runner ✔ (just K2 ones)
│   └── runner ✔ (just HC ones)
├── Cargo.lock ✔ (all because deps might have changed?)
├── Cargo.toml ✘ (covered by lock-file)
├── CHANGELOG.md ✘
├── CLAUDE.md ✘
├── dnas
│   └── <scenario-name> ✔ (just the scenario whose DNA changed)
├── durable_object_store/..
├── flake.lock ✘
├── flake.nix ✘
├── framework/.. ✔ (all)
├── happ_builder ✔ (just HC ones)
├── happs/.. ✘ (output dir I think)
├── influx/.. ✘
├── LICENSE ✘
├── logs ✘
├── lp-tool/.. ✘
├── Makefile ✘
├── nix
│   └── modules ✔ (HC ones, hApp builder might have changed)
├── node_modules ✘
├── nomad/.. ✘
├── README.md ✘
├── rust-toolchain.toml ✔ (all, rustc version might have changed)
├── scenarios
│   └── <scenario-name> ✔ (just the scenario whose DNA changed)
├── scenarios_common/.. ✔ (all)
├── scripts ✘
├── summariser ? (will this change how metrics are captured?)
├── summary-visualiser ✘
├── taplo.toml ✘
├── telegraf/.. ✘
├── threefold/.. ✘
└── zomes ✔ (all HC ones unless we can easily see which scenarios are effected)
```

### TODO:

- [ ] All code changes are reflected in docs, including module-level docs
- [ ] All new/edited/removed scenarios are reflected in summary visualiser tool (see [checklist](https://github.com/holochain/wind-tunnel/tree/main/summary-visualiser/README.md#maintenance))
- [ ] I ran the Nomad CI workflow successfully on my branch

_Note that all commits in a PR must follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0) before it can be merged, as these are used to generate the changelog_

### CODE REVIEW

File: .github/workflows/publish-release.yml [hunk 1]
```diff
@@ -12,3 +12,49 @@ jobs:
...
+          fi
+
+      - name: Delete old pre-releases
+        if: steps.check_release.outputs.skip == 'false'
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          gh release list --json tagName,isPrerelease \
+            --jq '.[] | select(.isPrerelease and (.tagName | startswith("main-"))) | .tagName' \
+            | xargs -r -I % gh release delete % --yes --cleanup-tag
```
[the remark this answers is not in the document]
Comment (cdunster) [reply .github/workflows/publish-release.yml:37]: There should only ever be one pre-release and the releases are returned in order so the one pre-release should always be within the 30 release limit, right?

File: .github/actions/nomad/action.yaml [hunk 2]
```diff
@@ -84,12 +88,30 @@ runs:
...
+          bindings/**
+          Cargo.lock
+          dnas/${{ steps.read-nomad-vars.outputs.scenario_name }}/**
+          framework/**
+          happ_builder/**
+          nix/modules/**
+          rust-toolchain.toml
+          scenarios/${{ steps.read-nomad-vars.outputs.scenario_name }}/**
+          scenarios_common/**
+          zomes/**
```
[the remark this answers is not in the document]
Comment (cdunster) [reply .github/actions/nomad/action.yaml:105]: An empty string is passed in this case which then fails the `== 'true'` check, acting the same as `'false'` which is the desired default behaviour.

File: .github/actions/nomad/action.yaml [hunk 3]
```diff
@@ -104,6 +126,32 @@ runs:
...
+        GH_TOKEN: ${{ github.token }}
+      run: |-
+        if [[ "$NEEDS_REBUILD" == 'true' || "$FORCE_REBUILD" == 'true' ]]; then
+          echo "Using newly built scenario from ${NEW_SCENARIO_URL}"
+          echo "scenario_download_url=${NEW_SCENARIO_URL}" >> "$GITHUB_OUTPUT"
+        else
+          TAG=$(gh api "/repos/${GITHUB_REPOSITORY}/releases?per_page=10" \
+            | jq -r --arg name "${SCENARIO_NAME}.zip" \
+              '.[] | select(any(.assets[]; .name == $name)) | .tag_name' \
+            | head -1)
```
Comment (cdunster) [inline .github/actions/nomad/action.yaml]: This searches for the scenario zip file in that latest 10 releases. This means that if a new release is created but the scenarios are still being built for it then it won't fail with a confusing error in Nomad due to the URL not existing. However, it does mean that if a new release is created during a run, then it is possible that the first bunch of scenario runs will use the old scenario bins from the latest release at the time and then rest will use the scenarios from the new release that was created during the run. I think this is a fine edge-case and it's up to the developer to make sure that the release they want to test with exists before running the Nomad jobs.

**Comment by cdunster** [comment]
I created a commit that changed the `app_install` scenario to test this and as expected the two jobs that use that scenario did the rebuild and the other jobs all took the scenario bins from the most recent tag.

You can see the results here: https://github.com/holochain/wind-tunnel/actions/runs/24671870481

**Review by ThetaSinner — DISMISSED** [review]

**Review by veeso — DISMISSED** [review]

**Review by veeso — APPROVED** [review]
